File: lab3/cpp/dh_protocol_test.cpp
#include <iostream>
#include <gtest/gtest.h>

#include "GaloisField.hpp"
#include "DHSetup.hpp"
#include "DHUser.hpp"

constexpr uint64_t PRIME = 1234567891;

using GF = GaloisField<PRIME>;

TEST(DHProtocol, ProtocolTest) {
    DHSetup<GF> setup;
    DHUser<GF> alice{setup};
    DHUser<GF> bob{setup};

    alice.setPublicKey(bob.getPublicKey());
    bob.setPublicKey(alice.getPublicKey());

    GF message1{123456789};
    GF message2{987654321};

    ASSERT_EQ(message1, alice.decrypt(bob.encrypt(message1)));
    ASSERT_EQ(message1, bob.decrypt(alice.encrypt(message1)));

    ASSERT_EQ(message2, alice.decrypt(bob.encrypt(message2)));
    ASSERT_EQ(message2, bob.decrypt(alice.encrypt(message2)));
}

TEST(DHProtocol, EncryptErrorKeyNotSet) {
    DHSetup<GF> setup;
    DHUser<GF> alice{setup};
    DHUser<GF> bob{setup};

    GF message{123456789};

    ASSERT_THROW(alice.encrypt(message), std::logic_error);
    ASSERT_THROW(bob.encrypt(message), std::logic_error);
}

TEST(DHProtocol, DecryptErrorKeyNotSet) {
    DHSetup<GF> setup;
    DHUser<GF> alice{setup};
    DHUser<GF> bob{setup};

    GF message{123456789};

    ASSERT_THROW(alice.decrypt(message), std::logic_error);
    ASSERT_THROW(bob.decrypt(message), std::logic_error);
}

int main() {
    ::testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: lab3/python/dh_protocol_test.py
import pytest

from galois_field import GaloisField
from dh_setup import DHSetup
from dh_user import DHUser

def test_dh_protocol():
    setup = DHSetup()
    alice = DHUser(setup)
    bob = DHUser(setup)

    alice.setPublicKey(bob.getPublicKey())
    bob.setPublicKey(alice.getPublicKey())

    message1 = GaloisField(123456789)
    message2 = GaloisField(987654321)

    assert alice.decrypt(bob.encrypt(message1)) == message1
    assert bob.decrypt(alice.encrypt(message1)) == message1

    assert alice.decrypt(bob.encrypt(message2)) == message2
    assert bob.decrypt(alice.encrypt(message2)) == message2


def test_dh_protocol_encrypt_key_not_set():
    setup = DHSetup()
    alice = DHUser(setup)

    message = GaloisField(123456789)

    with pytest.raises(Exception):
        alice.encrypt(message)

def test_dh_protocol_decrypt_key_not_set():
    setup = DHSetup()
    alice = DHUser(setup)

    message = GaloisField(123456789)

    with pytest.raises(Exception):
        alice.decrypt(message)
